### DISCUSSION TITLE
Piores praticas de programação

### DISCUSSION THREAD

**Discussion by munizart** [frontendbr/forum#863] [Perguntas e Respostas]
Olá pessoal!
Inspirado por artigos como "[Como não ser um bom front-end](https://medium.com/@kvnol/como-n%C3%A3o-ser-um-bom-front-end-bae2ce7215fb)" do @kvnol e afins, to pensando em escrever algo sobre más praticas de programação em `javascript`.

Então queria saber de vocês qual ou quais as piores praticas vocês já viram.
Vou listar algumas que **eu** considero ruim.
### Usar a keyoword [`with`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/with)
Como já avisa a MDN ([_vale a pena ler o artigo_](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/with)):
![Aviso da MDN sobre o uso de "with"](https://user-images.githubusercontent.com/4699689/34135466-118645b2-e448-11e7-8be7-801d7c0fcde8.png)
_Essa é realmente rara, para nossa alegria._

É uma _keyword_ que nem gosto de falar sobre, pra ninguém cair em tentação. O comportamento dela é bastante confuso (e cheio de inconsistências), mas basicamente permite você acessar propriedades de um objeto como se fossem varieis de escopo dentro de determinado bloco.
Exemplo:
```javascript
with (Math) {
  console.log(round(PI)) //3
}
```
A noticia boa é que o código abaixo te dá um _`SintaxError`_, já que o [modo estrito](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Strict_mode) bloqueia o uso de `with`.

```javascript
function strictFunction() {
  "use strict";
  with (Math) {
    console.log(round(PI)) //3
  }
}
```

### Usar um loop para mapear um array para outro
Essa é extremamente comum, a maior parte das vezes por desconhecimento da função [`Array.prototype.map`](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Array/map).
Eu especialmente não gosto porque todo _loop_ acaba criando mais uma variável no escopo pra levar em consideração (normalmente com um péssimo nome, tipo `i` ou `j`).
```javascript
var estrofes = [
  "brilha",
  "brilha",
  "estrelinha"
];
var refrao = [];
for (var i = 0; i < estrofe.length; i++) {
  refrao.push(estrofe[i].toUpperCase());
}
```
Com map ficaria:
```javascript
var estrofes = [
  "brilha",
  "brilha",
  "estrelinha"
];
var refrao = estrofes.map(function gritando(estrofe) {
  return estrofe.toUpperCase();
});
```

### Fallback com operador ternário
Isso:
```javascript
function Pessoa (nome) {
  this.nome = nome ? nome : "Maria";
}
```
Invés disso:
```javascript
function Pessoa (nome) {
  this.nome = nome || "Maria";
}
```

### Código em português
Como vocês viram em todos os exemplos _javascript_ é uma linguagem com _keywords_ em inglês e toda essa troca de contexto entre inglês/português no meio código realmente dificulta a leitura de códigos maiores.
Não me levem a mal, português é minha língua favorita. Mas assim como não gosto de falar palavras como "_blueberry_" ou "_hashtag_" no meio das frases em português, também não gosto de usar nome de variáveis e funções em pt-BR.

E pra vocês, quais seriam as piores praticas? discordam de algo que disse?

**Comment by kivervinicius** [comment]
Eu já conhecia o with em outras linguagens, para ser sincero nem sabia que esse "empestiado" existia no JS.

**Comment by kvnol** [comment]
Que massa que meu artigo te inspirou, @munizart. Faz o seguinte, como eu fiz no meu artigo, coloca ele no ar e vai fazendo updates assim que pegar novas más práticas.

**Comment by Sergioamjr** [comment]
O Eslint me ajuda muito pra evitar esses erros, uma coisa que eu uso com frequencia é o destructuring,
Ao invês de
```
const id = obj.key
```
Usar
```
const { key: id } = obj
```

Referência http://exploringjs.com/es6/ch_destructuring.html

**Comment by dionarodrigues** [comment]
No Browser Diet tem bastante dicas interessantes: https://browserdiet.com/pt/#js

**Comment by AmandaFonseca** [comment]
A pior pratica pelo menos a que mais me prejudica, é quando eu vou pesquisar algum erro no google e acabo clicando em links aleatórios  e no final do dia descubro que eu gastei a maior parte do meu dia lendo sobre a vida secreta dos golfinhos ....

**Comment by kivervinicius** [comment]
@AmandaFonseca tmj

**Comment by phsantiago** [comment]
### Não escrever testes

**Comment by romulo2735** [comment]
Ainda sou muito noob em JS, mas vcs estão me ajudando! hehe

**Comment by Pedro-Souza** [comment]
Quando escrever poste aqui o link
